plugins/home/src/components/VisitListener.tsx

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React, {
+  createContext,
+  useState,
+  useEffect,
+  useContext,
+  Dispatch,
+  SetStateAction,
+  ReactNode,
+} from 'react';
+
+import { useLocation } from 'react-router-dom';
+
+import { visitsApiRef } from '../api';
+import { useApi } from '@backstage/core-plugin-api';
+import { stringifyEntityRef } from '@backstage/catalog-model';
+
+/** @public */
+export type VisitListenerContextValue = {
+  doNotTrack: boolean;
+  setDoNotTrack: Dispatch<SetStateAction<boolean>>;
+};
+
+const defaultVisitListenerContext: VisitListenerContextValue = {
+  doNotTrack: false,
+  setDoNotTrack: () => {},
+};
+
+/** @public */
+export const VisitListenerContext = createContext<VisitListenerContextValue>(
+  defaultVisitListenerContext,
+);
+
+/**
+ * @public
+ * This function returns an implementation of toEntityRef which is responsible
+ * for receiving a pathname and maybe returning an entityRef compatible with the
+ * catalog-model.
+ * By default this function uses the url root "/catalog" and the
+ * stringifyEntityRef implementation from catalog-model.
+ * Example:
+ *   const toEntityRef = getToEntityRef();
+ *   toEntityRef(\{ pathname: "/catalog/default/component/playback-order" \})
+ *   // returns "component:default/playback-order"
+ */
+export const getToEntityRef =
+  ({
+    rootPath = 'catalog',
+    stringifyEntityRefImpl = stringifyEntityRef,
+  } = {}) =>
+  ({ pathname }: { pathname: string }): string | undefined => {
+    const regex = new RegExp(
+      `^\/${rootPath}\/(?<namespace>[^\/]+)\/(?<kind>[^\/]+)\/(?<name>[^\/]+)`,
+    );
+    const result = regex.exec(pathname);
+    if (!result || !result?.groups) return undefined;
+    const entity = {
+      namespace: result.groups.namespace,
+      kind: result.groups.kind,
+      name: result.groups.name,
+    };
+    return stringifyEntityRefImpl(entity);
+  };
+
+/**
+ * @public
+ * This function returns an implementation of visitName which is responsible
+ * for receiving a pathname and returning a string (name). The default
+ * implementation ignores the pathname and uses the document.title .
+ */
+export const getVisitName = (document: Document) => () => document.title;
+
+/**
+ * @public
+ * Component responsible for listening to location changes and calling
+ * the visitsApi to save visits.
+ */
+export const VisitListener = ({
```

Comment:
Non-blocking suggestion, could totally be addressed in a follow-up pull request:
Should this rely on the route tracker as the [Analytics plugin does](https://github.com/backstage/backstage/blob/5995614eb036ec9b9cec8fd40e41df8c11f28be3/packages/core-app-api/src/routing/RouteTracker.tsx#L133)? If this is always associated with a pathname (required attribute on Visit type), I wonder if visits could be saved by the router tracker out of box once there is a Visits API implementation available.
One benefit would be not requiring this extra code configuration from the adopter (manually wrap the app with the Visit context) and it might make it easy to use with declarative integration in case it requires less code.

Reply:
I love this idea since the first time you spoke to me about it. Thank you for leaving this registered here.
I would like to investigate this in a following PR because I need some investigation time. But 100% agree that if it works, it will be way better.